jetty-quic/quic-quiche/quic-quiche-jna/src/test/java/org/eclipse/jetty/quic/quiche/jna/LowLevelQuicheClientCertTest.java

```diff
@@ -0,0 +1,249 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.quic.quiche.jna;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+
+import org.eclipse.jetty.quic.quiche.PemExporter;
+import org.eclipse.jetty.quic.quiche.QuicheConfig;
+import org.eclipse.jetty.quic.quiche.QuicheConnection;
+import org.eclipse.jetty.toolchain.test.jupiter.WorkDir;
+import org.eclipse.jetty.toolchain.test.jupiter.WorkDirExtension;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import static org.eclipse.jetty.quic.quiche.Quiche.QUICHE_MIN_CLIENT_INITIAL_LEN;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.core.Is.is;
+
+@ExtendWith(WorkDirExtension.class)
+public class LowLevelQuicheClientCertTest
+{
+    public WorkDir workDir;
+
+    private final Collection<JnaQuicheConnection> connectionsToDisposeOf = new ArrayList<>();
+
+    private InetSocketAddress clientSocketAddress;
+    private InetSocketAddress serverSocketAddress;
+    private QuicheConfig clientQuicheConfig;
+    private QuicheConfig serverQuicheConfig;
+    private JnaQuicheConnection.TokenMinter tokenMinter;
+    private JnaQuicheConnection.TokenValidator tokenValidator;
+    private Certificate[] serverCertificateChain;
+
+    @BeforeEach
+    protected void setUp() throws Exception
+    {
+        clientSocketAddress = new InetSocketAddress("localhost", 9999);
+        serverSocketAddress = new InetSocketAddress("localhost", 8888);
```

Comment:
Hardcoded ports?

Reply:
Yes, but it doesn't matter as this test does not bind the ports nor even uses sockets.

Reply:
Quiche does not communicate with the network, you have to give it buffers that you read yourself and ask it for the buffers to send, a bit like `SSLEngine`. Look carefully, this test does not use sockets anywhere but Quiche still needs to know from which host/port the buffers came from which explains why there are `InetSocketAddress`es in this test to mock that information.

Reply:
Doh! 🤦🏼‍♂️